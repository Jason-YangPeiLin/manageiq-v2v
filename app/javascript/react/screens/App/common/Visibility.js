import React from 'react';
import PropTypes from 'prop-types';
import { noop } from 'patternfly-react';

// This component has no role on the page, and will be hidden from screen readers (role="presentation").
// Renders the children wrapped in visibility: hidden if the condition is met.
// Either way, the children take up space, visibility will not change the flow of the layout.
// The onKeyUp handler and role are used to satisfy eslint-plugin-jsx-a11y.
const Visibility = ({ children, hidden }) => (
  <span style={{ visibility: hidden ? 'hidden' : '' }} onKeyUp={noop} role="presentation">
    {children}
  </span>
);

Visibility.propTypes = {
  children: PropTypes.node,
  hidden: PropTypes.bool
};

export default Visibility;
