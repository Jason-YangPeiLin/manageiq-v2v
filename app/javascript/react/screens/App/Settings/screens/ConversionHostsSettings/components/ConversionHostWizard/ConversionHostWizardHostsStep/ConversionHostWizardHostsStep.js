import React from 'react';
import PropTypes from 'prop-types';
import { reduxForm, Field } from 'redux-form';
import { length } from 'redux-form-validators';
import { Form } from 'patternfly-react';
import { RHV, OPENSTACK } from '../../../../../../../../../common/constants';
import TypeAheadSelectField from '../../../../../../common/forms/TypeAheadSelectField';
import { stepIDs } from '../ConversionHostWizardConstants';

const ConversionHostWizardHostsStep = ({ selectedProviderType, selectedCluster }) => {
  let hostOptions = [];
  let emptyLabel = '';
  if (selectedProviderType === RHV) {
    hostOptions = selectedCluster.hosts;
    emptyLabel = __('No hosts available for the selected cluster.');
  }
  if (selectedProviderType === OPENSTACK) {
    hostOptions = selectedCluster.vms;
    emptyLabel = __('No hosts available for the selected project.');
  }
  return (
    <Form className="form-vertical">
      <Form.FormGroup controlId="host-selection">
        <Form.ControlLabel>{__('Hosts to configure as conversion hosts')}</Form.ControlLabel>
        <Field
          component={TypeAheadSelectField}
          name="hosts"
          controlId="host-selection"
          multiple
          clearButton
          options={hostOptions}
          labelKey="name"
          placeholder={__('Select one or more hosts...')}
          emptyLabel={hostOptions.length === 0 ? emptyLabel : __('No matches found.')}
          highlightOnlyResult
          selectHintOnEnter
          validate={[length({ min: 1 })]}
        />
      </Form.FormGroup>
    </Form>
  );
};

ConversionHostWizardHostsStep.propTypes = {
  selectedProviderType: PropTypes.string,
  selectedCluster: PropTypes.object
};

export default reduxForm({
  form: stepIDs.hostsStep,
  destroyOnUnmount: false,
  initialValues: { hosts: [] }
})(ConversionHostWizardHostsStep);
