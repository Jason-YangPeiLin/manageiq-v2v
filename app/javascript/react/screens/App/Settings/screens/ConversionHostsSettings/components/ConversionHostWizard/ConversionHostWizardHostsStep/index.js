import ConversionHostWizardHostsStep from './ConversionHostWizardHostsStep';

export default ConversionHostWizardHostsStep;
