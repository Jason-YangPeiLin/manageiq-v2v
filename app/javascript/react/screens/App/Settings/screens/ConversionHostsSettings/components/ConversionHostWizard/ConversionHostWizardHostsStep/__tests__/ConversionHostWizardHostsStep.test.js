import React from 'react';
import { shallow } from 'enzyme';
import { reducer as formReducer } from 'redux-form';

import { generateStore } from '../../../../../../../common/testReduxHelpers';
import ConversionHostWizardHostsStep from '../ConversionHostWizardHostsStep';
import { RHV, CONVERSION_HOST_TYPES, OPENSTACK } from '../../../../../../../../../../common/constants';
import { FINISHED } from '../../../../ConversionHostsSettingsConstants';

const mockConfiguredRhvHost = { id: '1', name: 'already-configured-host', type: CONVERSION_HOST_TYPES[RHV] };
const mockInProgressRhvHost = { id: '2', name: 'host-being-configured', type: CONVERSION_HOST_TYPES[RHV] };

const mockConfiguredOspVm = { id: '1', name: 'already-configured-vm', type: CONVERSION_HOST_TYPES[OPENSTACK] };
const mockInProgressOspVm = { id: '2', name: 'vm-being-configured', type: CONVERSION_HOST_TYPES[OPENSTACK] };

const mockRhvCluster = {
  mock: 'cluster',
  hosts: [
    mockConfiguredRhvHost,
    mockInProgressRhvHost,
    { id: '3', name: 'available-host-1', type: CONVERSION_HOST_TYPES[RHV] },
    { id: '4', name: 'available-host-2', type: CONVERSION_HOST_TYPES[RHV] }
  ]
};

const mockOspTenant = {
  mock: 'tenant',
  vms: [
    mockConfiguredOspVm,
    mockInProgressOspVm,
    { id: '3', name: 'available-vm-1', type: CONVERSION_HOST_TYPES[OPENSTACK] },
    { id: '4', name: 'available-vm-2', type: CONVERSION_HOST_TYPES[OPENSTACK] }
  ]
};

const mockConversionHosts = [
  {
    mock: 'rhvConversionHost',
    resource: { ...mockConfiguredRhvHost }
  },
  {
    mock: 'ospConversionHost',
    resource: { ...mockConfiguredOspVm }
  }
];

const mockTasksByResource = {
  [CONVERSION_HOST_TYPES[RHV]]: {
    '1': { enable: [{ mock: 'task', state: FINISHED }] },
    '2': { enable: [{ mock: 'task', state: 'Active' }] }
  },
  [CONVERSION_HOST_TYPES[OPENSTACK]]: {
    '1': { enable: [{ mock: 'task', state: FINISHED }] },
    '2': { enable: [{ mock: 'task', state: 'Active' }] }
  }
};

describe('conversion host wizard hosts step', () => {
  const store = generateStore({ form: formReducer }, {});

  const baseProps = {
    selectedProviderType: RHV,
    selectedCluster: mockRhvCluster,
    conversionHosts: mockConversionHosts,
    conversionHostTasksByResource: mockTasksByResource,
    store
  };

  const shallowDive = jsx => {
    const connectWrapper = shallow(jsx); // <Connect(Form(ConversionHostWizardHostsStep))>
    const formWrapper = connectWrapper.dive(); // <Form(ConversionHostWizardHostsStep)>
    const componentWrapper = formWrapper.dive(); // <ConversionHostWizardHostsStep>
    return componentWrapper.dive(); // shallow-rendered ConversionHostWizardHostsStep
  };

  it('renders the redux-form wrapper correctly', () => {
    const component = shallow(<ConversionHostWizardHostsStep {...baseProps} />);
    expect(component).toMatchSnapshot();
  });

  it('renders correctly in the RHV example case', () => {
    const component = shallowDive(<ConversionHostWizardHostsStep {...baseProps} />);
    expect(component).toMatchSnapshot();
  });

  it('filters out options for RHV hosts that are configured or being configured', () => {
    const component = shallowDive(<ConversionHostWizardHostsStep {...baseProps} />);
    const filteredHostOptions = component.find('Field[controlId="host-selection"]').props().options;
    expect(filteredHostOptions.find(option => option === mockConfiguredRhvHost)).toBeFalsy();
    expect(filteredHostOptions.find(option => option === mockInProgressRhvHost)).toBeFalsy();
  });

  it('renders correctly in the OSP example case', () => {
    const component = shallowDive(
      <ConversionHostWizardHostsStep {...baseProps} selectedProviderType={OPENSTACK} selectedCluster={mockOspTenant} />
    );
    expect(component).toMatchSnapshot();
  });

  it('filters out options for OSP hosts that are configured or being configured', () => {
    const component = shallowDive(
      <ConversionHostWizardHostsStep {...baseProps} selectedProviderType={OPENSTACK} selectedCluster={mockOspTenant} />
    );
    const filteredHostOptions = component.find('Field[controlId="host-selection"]').props().options;
    expect(filteredHostOptions.find(option => option === mockConfiguredOspVm)).toBeFalsy();
    expect(filteredHostOptions.find(option => option === mockInProgressOspVm)).toBeFalsy();
  });

  it('renders correctly when there are no hosts available', () => {
    const component = shallowDive(<ConversionHostWizardHostsStep {...baseProps} selectedCluster={{ hosts: [] }} />);
    expect(component).toMatchSnapshot();
  });
});
