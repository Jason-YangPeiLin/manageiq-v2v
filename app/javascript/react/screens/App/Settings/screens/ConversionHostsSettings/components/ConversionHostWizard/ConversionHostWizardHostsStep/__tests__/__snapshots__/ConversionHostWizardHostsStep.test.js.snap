// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`conversion host wizard hosts step renders correctly in the OSP example case 1`] = `
<Form
  bsClass="form"
  className="form-vertical"
  componentClass="form"
  horizontal={false}
  inline={false}
>
  <FormGroup
    bsClass="form-group"
    controlId="host-selection"
  >
    <ControlLabel
      bsClass="control-label"
      srOnly={false}
    >
      Hosts to configure as conversion hosts
    </ControlLabel>
    <Field
      clearButton={true}
      component={[Function]}
      controlId="host-selection"
      emptyLabel="No matches found."
      highlightOnlyResult={true}
      labelKey="name"
      multiple={true}
      name="hosts"
      options={
        Array [
          Object {
            "id": "3",
            "name": "available-vm-1",
            "type": "ManageIQ::Providers::Openstack::CloudManager::Vm",
          },
          Object {
            "id": "4",
            "name": "available-vm-2",
            "type": "ManageIQ::Providers::Openstack::CloudManager::Vm",
          },
        ]
      }
      placeholder="Select one or more hosts..."
      selectHintOnEnter={true}
      validate={
        Array [
          [Function],
        ]
      }
    />
  </FormGroup>
</Form>
`;

exports[`conversion host wizard hosts step renders correctly in the RHV example case 1`] = `
<Form
  bsClass="form"
  className="form-vertical"
  componentClass="form"
  horizontal={false}
  inline={false}
>
  <FormGroup
    bsClass="form-group"
    controlId="host-selection"
  >
    <ControlLabel
      bsClass="control-label"
      srOnly={false}
    >
      Hosts to configure as conversion hosts
    </ControlLabel>
    <Field
      clearButton={true}
      component={[Function]}
      controlId="host-selection"
      emptyLabel="No matches found."
      highlightOnlyResult={true}
      labelKey="name"
      multiple={true}
      name="hosts"
      options={
        Array [
          Object {
            "id": "3",
            "name": "available-host-1",
            "type": "ManageIQ::Providers::Redhat::InfraManager::Host",
          },
          Object {
            "id": "4",
            "name": "available-host-2",
            "type": "ManageIQ::Providers::Redhat::InfraManager::Host",
          },
        ]
      }
      placeholder="Select one or more hosts..."
      selectHintOnEnter={true}
      validate={
        Array [
          [Function],
        ]
      }
    />
  </FormGroup>
</Form>
`;

exports[`conversion host wizard hosts step renders correctly when there are no hosts available 1`] = `
<Form
  bsClass="form"
  className="form-vertical"
  componentClass="form"
  horizontal={false}
  inline={false}
>
  <FormGroup
    bsClass="form-group"
    controlId="host-selection"
  >
    <ControlLabel
      bsClass="control-label"
      srOnly={false}
    >
      Hosts to configure as conversion hosts
    </ControlLabel>
    <Field
      clearButton={true}
      component={[Function]}
      controlId="host-selection"
      emptyLabel="No hosts available for the selected cluster."
      highlightOnlyResult={true}
      labelKey="name"
      multiple={true}
      name="hosts"
      options={Array []}
      placeholder="Select one or more hosts..."
      selectHintOnEnter={true}
      validate={
        Array [
          [Function],
        ]
      }
    />
  </FormGroup>
</Form>
`;

exports[`conversion host wizard hosts step renders the redux-form wrapper correctly 1`] = `
<Connect(Form(ConversionHostWizardHostsStep))
  conversionHostTasksByResource={
    Object {
      "ManageIQ::Providers::Openstack::CloudManager::Vm": Object {
        "1": Object {
          "enable": Array [
            Object {
              "mock": "task",
              "state": "Finished",
            },
          ],
        },
        "2": Object {
          "enable": Array [
            Object {
              "mock": "task",
              "state": "Active",
            },
          ],
        },
      },
      "ManageIQ::Providers::Redhat::InfraManager::Host": Object {
        "1": Object {
          "enable": Array [
            Object {
              "mock": "task",
              "state": "Finished",
            },
          ],
        },
        "2": Object {
          "enable": Array [
            Object {
              "mock": "task",
              "state": "Active",
            },
          ],
        },
      },
    }
  }
  conversionHosts={
    Array [
      Object {
        "mock": "rhvConversionHost",
        "resource": Object {
          "id": "1",
          "name": "already-configured-host",
          "type": "ManageIQ::Providers::Redhat::InfraManager::Host",
        },
      },
      Object {
        "mock": "ospConversionHost",
        "resource": Object {
          "id": "1",
          "name": "already-configured-vm",
          "type": "ManageIQ::Providers::Openstack::CloudManager::Vm",
        },
      },
    ]
  }
  destroyOnUnmount={false}
  enableReinitialize={false}
  forceUnregisterOnUnmount={false}
  form="conversionHostWizardHostsStep"
  getFormState={[Function]}
  initialValues={
    Object {
      "hosts": Array [],
    }
  }
  keepDirtyOnReinitialize={false}
  persistentSubmitErrors={false}
  pure={true}
  selectedCluster={
    Object {
      "hosts": Array [
        Object {
          "id": "1",
          "name": "already-configured-host",
          "type": "ManageIQ::Providers::Redhat::InfraManager::Host",
        },
        Object {
          "id": "2",
          "name": "host-being-configured",
          "type": "ManageIQ::Providers::Redhat::InfraManager::Host",
        },
        Object {
          "id": "3",
          "name": "available-host-1",
          "type": "ManageIQ::Providers::Redhat::InfraManager::Host",
        },
        Object {
          "id": "4",
          "name": "available-host-2",
          "type": "ManageIQ::Providers::Redhat::InfraManager::Host",
        },
      ],
      "mock": "cluster",
    }
  }
  selectedProviderType="rhevm"
  shouldAsyncValidate={[Function]}
  shouldError={[Function]}
  shouldValidate={[Function]}
  shouldWarn={[Function]}
  store={
    Object {
      "dispatch": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
      Symbol(observable): [Function],
    }
  }
  touchOnBlur={true}
  touchOnChange={false}
  updateUnregisteredFields={false}
/>
`;
